PatternsTools namespace.

luca-heltai opened the pull request:
Superseeds #4286.

This is a full restructuring of the code I initially proposed. Implementation is header only, ~~therefore the `add_parameter` function is actually in the same namespace of the `Convert` function~~  and the final functionality is in ParameterHandler::add_parameter.

The boost::any solution was not flexible enough, so I opted for a lot of SFINAE on a new struct, that is capable to convert to/from and to generate a default pattern.

The hard worker of this PR is `PatternsTools::Convert<T>` and its three static member functions:

~~~
Pattern::Tools::Convert<T>::to_pattern()
Pattern::Tools::Convert<T>::to_value()
Pattern::Tools::Convert<T>::to_string()
~~~

These changes are transparent to `ParameterHandler` (except for a few getter functions) and allow all possible combinations of types I could come up with, including all stl container types, and some map types. 

I have not tested all cases, and certainly some of them won't work, but it's easy enough to extend and to fix it. 

This PR goes in the direction of addressing most of the concerns that @bangerth had on my previous one. 

~~I'd keep the `add_parameter` function where it is now, instead of making it a member function of `prm`, since this header only file includes a lot of std containers and map types. I'm not sure how bad this would be in `parameter_handler.h`~~

luca-heltai commented:
I'll keep adding tests, but I wanted some feedback before I spend too much time on this...

bangerth commented:
There's currently this compile failure:
```

all-headers/base/patterns_tools.h.debug: BUILD failed. Output:
Generating base-patterns_tools.h.debug_interrupt_guard.cc
Scanning dependencies of target base-patterns_tools.h.debug
Building CXX object CMakeFiles/base-patterns_tools.h.debug.dir/test_header.cc.o
In file included from /home/bob/source/tests/all-headers/test_header.cc:17:
/home/bob/source/include/deal.II/base/patterns_tools.h:316:73: error: no template named 'decay_t' in namespace 'std'; did you mean 'decay'?
    static constexpr bool const value = internal::is_stl_container<std::decay_t<T>>::value;
                                                                   ~~~~~^~~~~~~
                                                                        decay
/usr/bin/../lib/gcc/x86_64-linux-gnu/4.8/../../../../include/c++/4.8/type_traits:1725:11: note: 'decay' declared here
    class decay 
          ^
In file included from /home/bob/source/tests/all-headers/test_header.cc:17:
/home/bob/source/include/deal.II/base/patterns_tools.h:322:67: error: no template named 'decay_t' in namespace 'std'; did you mean 'decay'?
    static constexpr bool const value = internal::is_stl_map<std::decay_t<T>>::value;
                                                             ~~~~~^~~~~~~
                                                                  decay
/usr/bin/../lib/gcc/x86_64-linux-gnu/4.8/../../../../include/c++/4.8/type_traits:1725:11: note: 'decay' declared here
    class decay 
          ^
2 errors generated.
```

bangerth commented on include/deal.II/base/patterns_tools.h:
```diff
@@ -0,0 +1,682 @@
+     * class, or an elementary type.
+     *
+     * Elementary types are not compatible with Patterns::List, but
+     * non elementary types, like Point<dim>(), or std::complex<double>, are
```
Point<dim>() -> Point<dim>

bangerth commented on include/deal.II/base/patterns_tools.h:
```diff
@@ -0,0 +1,682 @@
+      if (std::is_integral<T>::value)
+        return std_cxx14::make_unique<Integer>(std::numeric_limits<T>::min(), std::numeric_limits<T>::max());
+      else if (std::is_floating_point<T>::value)
+        return std_cxx14::make_unique<Double>(-std::numeric_limits<T>::max(), std::numeric_limits<T>::max());
```
I think it would be useful to omit the min/max values here so that we just default to what these classes want to use, without trying to duplicate their default constructor values.

luca-heltai replied:
This is actually needed for manifold_ids, boundary_ids, unsigned int, etc. With this constructor, I make sure that boundary_ids are not treated as integers, and that unsigned int are not treated as ints.

bangerth replied:
Hm, I see. OK then.

bangerth commented on include/deal.II/base/patterns_tools.h:
```diff
@@ -0,0 +1,682 @@
+      else
+        {
+          AssertThrow(false, ExcNotImplemented());
+          return std::unique_ptr<PatternBase>();
```
You can avoid this case if you explicitly copy the two conditions -- namely that the type `T` must either satisfy `is_integral` or `is_floating_point` into the `enable_if` condition of the surrounding class.

luca-heltai replied:
Oh, actually it is not necessary. :)

~~~

template <class _Tp> struct _LIBCPP_TYPE_VIS_ONLY is_arithmetic
    : public integral_constant<bool, is_integral<_Tp>::value      ||
                                     is_floating_point<_Tp>::value> {};
~~~

bangerth replied:
Oh, nice.

bangerth commented on include/deal.II/base/patterns_tools.h:
```diff
@@ -0,0 +1,682 @@
+    {
+      std::string str;
+      if (std::is_same<T, unsigned char>())
+        str = std::to_string((int)value);
```
what about `signed char`?

bangerth commented on include/deal.II/base/patterns_tools.h:
```diff
@@ -0,0 +1,682 @@
+      AssertThrow(p->match(s), ExcNoMatch(s, *p));
+      std::istringstream is(s);
+      T value;
+      if (std::is_same<T,unsigned char>::value)
```
same here for `signed char`

bangerth commented on include/deal.II/base/patterns_tools.h:
```diff
@@ -0,0 +1,682 @@
+        }
+      else
+        is >> value;
+      AssertThrow(!is.fail(), ExcMessage("Failed to convert from \"" + s + "\" to the type \""
```
There is a tricky case here that you'll also find discussed in the `Utilities::string_to_int` functions and similar, namely what happens if someone passes "123 abc" to the function. Reading from this string yields an integer 123 alright, but the space terminates the read from the string although there is more to come. If you really want to cover all possibly wrong cases, you'll have to ensure there is nothing coming after the initial number.

luca-heltai replied:
This is checked by the `match(s)` at the beginning.

bangerth replied:
Good point. Can you add a comment to this effect, referencing the case I mentioned? Since it wasn't clear to me, it may not be clear to the next reader either.

bangerth commented on include/deal.II/base/patterns_tools.h:
```diff
@@ -0,0 +1,682 @@
+  namespace internal
+  {
+    const std::array<std::string, 5> default_list_separator {{" ",  ","  ,  ";"  ,  "|"  ,   "%"}};
+    const std::array<std::string, 5> default_map_separator {{" ",  ":"  ,  "="  ,  "@"  ,   "#"}};
```
can you document in particular what element 0 indicates here?

bangerth commented on include/deal.II/base/patterns_tools.h:
```diff
@@ -0,0 +1,682 @@
+    //specialize a type for all of the STL containers and maps
+    template <typename T>       struct is_stl_container:std::false_type {};
+    template <typename T, std::size_t N> struct is_stl_container<std::array    <T,N>>    :std::true_type {};
+    template <typename... Args> struct is_stl_container<std::vector            <Args...>>:std::true_type {};
```
maybe put a space around the `:` that denotes derivation here, to ease readability? same below, of course.

bangerth commented on include/deal.II/base/patterns_tools.h:
```diff
@@ -0,0 +1,682 @@
+    {
+      return std_cxx14::make_unique<List>(*Convert<typename T::value_type>::to_pattern(),
+                                          std::numeric_limits<typename T::size_type>::min(),
+                                          std::numeric_limits<typename T::size_type>::max(),
```
Shouldn't the second argument to this call just be zero?

bangerth replied:
And don't you get an overflow for the third argument? `Patterns::List` only takes `unsigned int`s in these two places.

bangerth commented on include/deal.II/base/patterns_tools.h:
```diff
@@ -0,0 +1,682 @@
+      T t;
+
+      auto v = Utilities::split_string_list(s,p->get_separator());
+      for (auto str : v)
```
use `const auto &str` for efficiency

bangerth commented on include/deal.II/base/patterns_tools.h:
```diff
@@ -0,0 +1,682 @@
+      T t;
+
+      auto v = Utilities::split_string_list(s,p->get_separator());
+      for (auto str : v)
```
`const auto &`

bangerth commented on include/deal.II/base/patterns_tools.h:
```diff
@@ -0,0 +1,682 @@
+
+      auto v = Utilities::split_string_list(s,p->get_separator());
+      unsigned int i=0;
+      for (auto str : v)
```
same here and probably everywhere else

bangerth commented on include/deal.II/base/patterns_tools.h:
```diff
@@ -0,0 +1,682 @@
+
+
+  // Strings
+  template<>
```
you have a mix of styles `template <...>` and `template<...>` (with or without space). I had a patch a couple of weeks ago that settles on the first style (with space). Can you convert all of your cases to the one with a space?

masterleinad commented on include/deal.II/base/patterns_tools.h:
```diff
@@ -0,0 +1,682 @@
+  //type trait to use the implementation type traits as well as decay the type
+  template <typename T> struct is_stl_container
+  {
+    static constexpr bool const value = internal::is_stl_container<std::decay_t<T>>::value;
```
`decay_t` is only available from C++14 on. Using `std::decay<T>::type` instead hopefully doesn't hurt to much :wink:

luca-heltai commented on include/deal.II/base/patterns_tools.h:
```diff
@@ -0,0 +1,682 @@
+#include <deal.II/base/parameter_handler.h>
+#include <deal.II/base/std_cxx14/memory.h>
+
+#include <boost/core/demangle.hpp>
```
Yes, very nicely. :) 

With STL containers it is a little verbose... But it's useful to debug what goes wrong...

luca-heltai commented on include/deal.II/base/patterns_tools.h:
```diff
@@ -0,0 +1,682 @@
+ *
+ * @code
+ * default_list_separator{{" ",  ","  ,  ";"  ,  "|"  ,   "%"}};
+ * default_map_separator {{" ",  ":"  ,  "="  ,  "@"  ,   "#"}};
```
Rank zero is not a List. I'll remove it, and access the list with `rank-1`.

luca-heltai commented on include/deal.II/base/patterns_tools.h:
```diff
@@ -0,0 +1,682 @@
+   * the given type, and to convert from a string to the given type and viceversa.
+   */
+  template<class T, class Enable=void>
+  struct Convert
```
I wanted to document what the three members do. Without a body, I'd have to provide the documentation below.

bangerth replied:
Yes, that occurred to me as well later on. I think this makes sense even though it looks a bit funny.

Maybe add something like "While the current function (in the general Convert template) is deleted, it is implemented and available in the specializations of the Convert class template for particular kinds of template arguments @p T."

luca-heltai commented:
> The #include list of this header file is really monstrous. It's a bit of a shame that we can't put this functionality into ParameterHandler itself right away. Or should we just ignore the issue of compile time?

The actual compile time is not too bad in the end. Let me continue in this way until the end (this allows me to make changes without recompiling deal.II, while inserting it in `parameter_handler.h` requires me to recompile the library for every little change, and believe me, with this I need many...).

I'll move the single function `add_parameter` to `parameter_handler`.

In the end it may even make sense to just put the Convert struct in the `Patterns` namespace. What do you think? Maybe separating `patterns.h` from `parameter_handler.h`, so we don't clutter `parameter_handler.h` with additional 1.000 lines of code...

If you all agree, I'll do this as a last commit to this patch before merging.

bangerth commented:
Yes, I think that's a good plan. I agree that breaking `patterns.h` (and `patterns.cc`) out of the existing file makes sense.

luca-heltai commented:
Should I keep to `PatternTools` namespace, or simply put everything in the `Patterns` namespace, after splitting the files?

bangerth commented:
Your choice. Namespace `Patterns` is not overly large, so there is space. On the other hand, everything that's in there is very similar, so a sub-namespace might make sense.

luca-heltai commented:
This is now ready. I moved everything into `patterns.h`, and removed `pattern_tools.h`.  The new struct is in `Patterns::Tools::Convert`. `add_parameter` has been added directly to `parameter_handler.h`. Compilation speed seems fine to me.

tjhei commented on doc/news/changes/minor/20170720LucaHeltai:
```diff
@@ -0,0 +1,11 @@
+New: The new namespace Patterns::Tools contain now some 
```
"... contains utilities that can be used ..."

tjhei commented on doc/news/changes/minor/20170720LucaHeltai:
```diff
@@ -0,0 +1,11 @@
+New: The new namespace Patterns::Tools contain now some 
+utilities that can be used to Convert from complicated types
```
lowercase "convert"

tjhei commented on include/deal.II/base/patterns.h:
```diff
@@ -0,0 +1,1681 @@
+        else if (std::is_floating_point<T>::value)
+          return std_cxx14::make_unique<Patterns::Double>(
+                   -std::numeric_limits<T>::max(), std::numeric_limits<T>::max());
+      }
```
does this give a warning of a missing return value?

luca-heltai replied:
no, because all possible cases are caught (is_aritmethic = is_integral || is_floating_point)

tjhei commented on include/deal.II/base/patterns.h:
```diff
@@ -0,0 +1,1681 @@
+      const std::array<std::string, 4> default_map_separator {{":"  ,  "="  ,  "@"  ,   "#"}};
+
+      //specialize a type for all of the STL containers and maps
+      template <typename T>       struct is_stl_container : std::false_type {};
```
I would have expected that a map is also an ``stl_container``. Is the naming a bit unfortunate?

luca-heltai replied:
Yes. Changed it.

tjhei commented on include/deal.II/base/patterns.h:
```diff
@@ -0,0 +1,1681 @@
+        auto p = dynamic_cast<const Patterns::List *>(pattern.get());
+        AssertThrow(
+          p,
+          ExcMessage(
```
weird newline.

bangerth commented on doc/news/changes/minor/20170720LucaHeltai:
```diff
@@ -0,0 +1,11 @@
+New: The new namespace Patterns::Tools contains 
```
by the way, I think this ought to go into `major/`

luca-heltai commented:
I rebased and removed some smaller commits. I kept most major features in separate commits. This is a large patch, and I think it is better to keep it separated in small-ish commits. 

For me, this is ready.

bangerth commented:
Let's do it!
